packages/apollo-server-core/src/runHttpQuery.ts

```diff
@@ -127,15 +129,13 @@ export async function runHttpQuery(
     // debug. Therefore, we need to do some unnatural things, such
     // as use NODE_ENV to determine the debug settings
     return throwHttpGraphQLError(500, [e], {
-      debug: debugFromNodeEnv(process.env.NODE_ENV),
+      debug: debugFromNodeEnv(NODE_ENV),
     });
   }
 
   if (options.debug === undefined) {
     const nodeEnv =
-      '__testing_nodeEnv__' in options
-        ? options.__testing_nodeEnv__
-        : process.env.NODE_ENV;
+      '__testing_nodeEnv__' in options ? options.__testing_nodeEnv__ : NODE_ENV;
     options.debug = debugFromNodeEnv(nodeEnv);
   }
```

Comment:
Here I would like to be sensitive to runtime NODE_ENV changes.  We actually already capture NODE_ENV once in the ApolloServer constructor and pass that through to various ways. Unfortunately for historical reasons much of the core processing in this package happens technically outside of the ApolloServer class even though it is only usable when called from ApolloServer, so this file doesn't have access to that.

That said, I basically implemented this already but just as a "testing" hack.  What I'd suggest is taking the `__testing_nodeEnv__` fields of `Config` and `GraphQLServerOptions` and renaming them to `nodeEnv`. Then in the place in `packages/apollo-server-core/src/ApolloServer.ts` where we do `const nodeEnv =` we should set that on `this.requestOptions.nodeEnv` instead.  That should lead to `options.nodeEnv` on this line in runHttpQuery being equal to the env var's value (or a manual override) already, so no need to read the env var in this file.

Is this something you can try?

Reply:
I've spent a few days trying to do this, and I'm not sure what to do.

The logic differs on the _existence_ of `nodeEnv`, not necessarily its value. More specifically, providing `nodeEnv: undefined` is _not_ the same thing as omitting it altogether.

What I would expect to happen is that if you don't provide `nodeEnv`, ApolloServer would default to the `NODE_ENV` environment variable. In other words, if `nodeEnv` is `undefined`, then default to `NODE_ENV`.

This doesn't work with the current implementation, at least as per the tests. The current implementation will simply check if you _attempt_ to set it, even if it's value is `undefined`; as long as it's there, it's considered an override.

So what is happening is that the tests fail, because when `nodeEnv` is not provided, it defaults to `NODE_ENV`, and when `NODE_ENV` is `test`, `debug` is false, and you don't get stacktraces back in `errors`.

The old implementation would avoid defaulting to `test` because `__testing_nodeEnv__` was simply there, even though it was `undefined`.

I'm not sure how to accommodate this.

Reply:
That being said, I don't think it's possible to get this to not refer `NODE_ENV` in some way anyways, because `resolveGraphqlOptions` is what is supposed to give you the `nodeEnv` option, but _that_ can error out and put you in a state where you need to know the node environment, but have no ability to get `nodeEnv`, and thus have to default to `NODE_ENV` anyways.

This would also introduce a quirk that even if you provided `nodeEnv`, it will be ignored depending on how the thing errors out, which sounds like a foot gun to me.

Reply:
When `__testing_nodeEnv__` was only being used in a few tests, making `undefined` distinct from "unset" seemed like a reasonable choice. It sounds like this is more of a burden when it's an actual part of the API.

How about we just change that?  So we could make `nodeEnv: undefined` and "`nodeEnv` unset" be the same thing (use the actual environment variable if it's set), and we can use `nodeEnv: ''` to mean "seriously, don't be sensitive to the actual process environment, act as if the node env is unspecified".  (Technically in the Unix model, there is a distinction between unset environment variables and environment variables that are set to the empty string, but it's not great to treat them differently and our code that reads this doesn't, so this should be fine.)

Does that resolve all the concerns you have here?

Reply:
Got it!